merkle-tree/src/merkle_tree.rs

```diff
@@ -0,0 +1,243 @@
+            for i in 0..level_len {
+                let prev_level_idx = 2 * i;
+                let lsib = &mt.nodes[prev_level_start + prev_level_idx];
+                let rsib = if prev_level_idx + 1 < prev_level_len {
+                    &mt.nodes[prev_level_start + prev_level_idx + 1]
+                } else {
+                    // Duplicate last entry if the level length is odd
+                    &mt.nodes[prev_level_start + prev_level_idx]
+                };
+
+                let hash = hashv(&[lsib.as_ref(), rsib.as_ref()]);
+                mt.nodes.push(hash);
+            }
+            prev_level_start = level_start;
+            prev_level_len = level_len;
+            level_start += level_len;
+            level_len = MerkleTree::next_level_len(level_len);
+        }
+
+        mt
+    }
+
+    pub fn get_root(&self) -> Option<&Hash> {
+        self.nodes.iter().last()
+    }
+
+    pub fn find_path(&self, index: usize) -> Option<Proof> {
+        if index >= self.leaf_count {
+            return None;
+        }
+
+        let mut level_len = self.leaf_count;
+        let mut level_start = 0;
+        let mut path = Proof::default();
+        let mut node_index = index;
+        let mut lsib = None;
+        let mut rsib = None;
+        while level_len > 0 {
+            let level = &self.nodes[level_start..(level_start + level_len)];
+
+            let target = &level[node_index];
+            if lsib != None || rsib != None {
+                path.push(ProofEntry::new(target, lsib, rsib));
+            }
+            if node_index % 2 == 0 {
+                lsib = None;
+                rsib = if node_index + 1 < level.len() {
+                    Some(&level[node_index + 1])
+                } else {
+                    Some(&level[node_index])
+                };
+            } else {
+                lsib = Some(&level[node_index - 1]);
+                rsib = None;
+            }
+            node_index /= 2;
+
+            level_start += level_len;
+            level_len = MerkleTree::next_level_len(level_len);
+        }
+        Some(path)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use hex;
+
+    const TEST: &'static [&'static [u8]] = &[
+        b"my", b"very", b"eager", b"mother", b"just", b"served", b"us", b"nine", b"pizzas",
+        b"make", b"prime",
+    ];
+    const BAD: &'static [&'static [u8]] = &[b"bad", b"missing", b"false"];
+
+    #[test]
+    fn test_tree_from_empty() {
+        let mt = MerkleTree::new(&[]);
+        assert_eq!(mt.get_root(), None);
+    }
+
+    #[test]
+    fn test_tree_from_one() {
+        let input = b"test";
+        let mt = MerkleTree::new(&[input]);
+        let expected = hash(input);
+        assert_eq!(mt.get_root(), Some(&expected));
+    }
+
+    #[test]
+    fn test_tree_from_many() {
+        let mt = MerkleTree::new(TEST);
+        let bytes = hex::decode("7e6791d2b9a6338e1446ad4776f267b97e7e4ed968ae2592cfd9c1607bd7dbbb")
```

Comment:
might be obvious to folks coming later, but maybe leave a note that this is a golden and cover conditions under which it would need to be updated

Reply:
https://github.com/solana-labs/solana/pull/4749/commits/c543ec6f4607b457d8e64a68379527273a647c45